python/cudf/cudf/core/series.py

```diff
@@ -216,17 +216,25 @@ def __setitem__(self, key, value):
         if (
             not isinstance(
                 self._frame._column.dtype,
-                (cudf.core.dtypes.DecimalDtype, cudf.CategoricalDtype),
+                (
+                    # Casting only for string and non-decimal numeric
+                    # columns
+                    cudf.core.dtypes.DecimalDtype,
+                    cudf.CategoricalDtype,
+                    cudf.ListDtype,
+                    cudf.StructDtype,
+                ),
             )
             and hasattr(value, "dtype")
             and _is_non_decimal_numeric_dtype(value.dtype)
         ):
             # normalize types if necessary:
-            if not is_integer(key):
-                to_dtype = np.result_type(
-                    value.dtype, self._frame._column.dtype
-                )
-                value = value.astype(to_dtype)
+            # In contrast to Column.__setitem__ (which downcasts the value to
+            # the dtype of the column) here we upcast the series to the
+            # larger data type mimicing pandas
+            to_dtype = np.result_type(value.dtype, self._frame._column.dtype)
+            value = value.astype(to_dtype)
+            if to_dtype != self._frame._column.dtype:
                 self._frame._column._mimic_inplace(
                     self._frame._column.astype(to_dtype), inplace=True
```

Comment:
The structure of this logic makes it seem as though `astype` lacks a fast path for doing nothing / returning `self` if the type matches the current type.

I would think we’d like to write `col = col.astype(x)` without this kind of conditional, and hope that it does as little work as possible. I might be overlooking some aspect of the “mimic inplace” though.

Reply:
`col.astype(col.dtype)` doesn't have a fast path (easy to add), `_mimic_inplace` is a strange beast: you would think that `col._mimic_inplace` would change `col`, but this is not true unless you explicitly pass `inplace=True`.

I think we can't just say `self._frame._column = self._frame._column.astype(to_dtype)` because there might be other handles to `_column` that would not be appropriately updated. e.g. what should this do:

```python
sr = cudf.Series([1,2,3])
shallow = sr.copy(deep=False)
sr[0] = 10
assert shallow[0] == 10 # should be true
# but if in `__setitem__` we do sr._column = sr._column.astype(...) it wil be false
```

One should be able to fast path `_mimic_inplace` for the case that `self is other` though.

Reply:
Added fastpath for astype, I haven't done mimic inplace yet though.